File: sdks/python/mindclade_internal_sdk/inference.py
"""Generated-type-only inference façade with resumable typed streaming."""

from __future__ import annotations

import asyncio
import threading
import time
from typing import cast

from google.protobuf.message import Message
from google.protobuf.timestamp_pb2 import Timestamp
from mindclade.inference.v1 import (
    inference_request_pb2,
    inference_result_pb2,
    inference_stream_pb2,
)
from mindclade.internal.inference.v1 import inference_service_pb2
from mindclade.operation.v1 import operation_pb2

from ._invocation import AsyncInvoker, SyncInvoker, canonical_digest, command_context
from ._raw import AsyncWithRawResponse, WithRawResponse, streaming_method
from ._validation import required_response_message, required_text
from ._watch import AsyncWatchStream, WatchSpec, WatchStream, watch_budget
from .calls import CallOptions, PreparedCall, prepare_call
from .errors import CancelledError, ProtocolError, UnavailableError
from .transport import (
    COMMIT_INFERENCE_RESULT,
    GET_INFERENCE_REQUEST,
    GET_INFERENCE_RESULT,
    SUBMIT_INFERENCE,
    WATCH_INFERENCE,
)


def _timestamp_after(seconds: float) -> Timestamp:
    value = Timestamp()
    value.FromNanoseconds(time.time_ns() + int(seconds * 1_000_000_000))
    return value


def _mutation_call(
    options: CallOptions | None,
    existing_key: str,
    default_timeout: float,
) -> PreparedCall:
    selected = options
    if selected is None and existing_key:
        selected = CallOptions(idempotency_key=existing_key)
    return prepare_call(selected, default_timeout=default_timeout, require_idempotency=True)


def _materialize_request(
    invoker: SyncInvoker | AsyncInvoker,
    request: inference_request_pb2.InferenceRequest,
    options: CallOptions | None,
) -> tuple[inference_request_pb2.InferenceRequest, PreparedCall]:
    value = inference_request_pb2.InferenceRequest()
    value.CopyFrom(request)
    existing_key = value.context.idempotency_key if value.HasField("context") else ""
    value.ClearField("context")
    value.tenant_id = invoker.config.tenant_id
    value.project_id = invoker.config.project_id
    required_text("inference request name", value.name)
    call = _mutation_call(options, existing_key, invoker.config.default_timeout)
    value.context.CopyFrom(
        command_context(invoker.config, call, request_digest=canonical_digest(value))
    )
    return value, call


def _materialize_commit(
    invoker: SyncInvoker | AsyncInvoker,
    command: inference_service_pb2.CommitInferenceResultRequest,
    options: CallOptions | None,
) -> tuple[inference_service_pb2.CommitInferenceResultRequest, PreparedCall]:
    value = inference_service_pb2.CommitInferenceResultRequest()
    value.CopyFrom(command)
    existing_key = value.context.idempotency_key if value.HasField("context") else ""
    value.ClearField("context")
    if not value.HasField("inference_request") or not value.HasField("fence"):
        raise ValueError("inference request reference and lease fence are required")
    if not value.HasField("result") or not value.request_digest:
        raise ValueError("inference result and request digest are required")
    call = _mutation_call(options, existing_key, invoker.config.default_timeout)
    value.context.CopyFrom(
        command_context(invoker.config, call, request_digest=canonical_digest(value))
    )
    return value, call


def _copied_cursor(
    cursor: inference_stream_pb2.InferenceStreamCursor | None,
) -> inference_stream_pb2.InferenceStreamCursor | None:
    """Copy a caller's cursor so the watch cannot mutate what it was handed."""

    if cursor is None:
        return None
    copy = inference_stream_pb2.InferenceStreamCursor()
    copy.CopyFrom(cursor)
    return copy


def _watch_request(
    operation_name: str,
    cursor: inference_stream_pb2.InferenceStreamCursor | None,
    remaining: float,
) -> inference_service_pb2.WatchInferenceRequest:
    request = inference_service_pb2.WatchInferenceRequest(
        operation_name=required_text("inference operation name", operation_name),
        deadline=_timestamp_after(remaining),
    )
    if cursor is not None:
        request.cursor.CopyFrom(cursor)
    return request


def _accept_message(
    response: inference_service_pb2.WatchInferenceResponse,
    cursor: inference_stream_pb2.InferenceStreamCursor | None,
) -> tuple[inference_stream_pb2.InferenceStreamMessage, inference_stream_pb2.InferenceStreamCursor]:
    message = required_response_message(
        response,
        "message",
        inference_stream_pb2.InferenceStreamMessage,
        label="inference watch",
    )
    required_text("inference request name", message.request_name)
    required_text("inference resume token", message.resume_token)
    if message.sequence <= 0 or message.WhichOneof("update") is None:
        raise ProtocolError("inference watch returned an incomplete message")
    after = cursor.after_sequence if cursor is not None else 0
    if message.WhichOneof("update") == "heartbeat":
        if (
            cursor is None
            or message.request_name != cursor.request_name
            or message.sequence != cursor.after_sequence
            or message.resume_token != cursor.resume_token
        ):
            raise ProtocolError("inference heartbeat is not bound to durable cursor")
        result = inference_stream_pb2.InferenceStreamCursor()
        result.CopyFrom(cursor)
        return message, result
    if cursor is not None and message.request_name != cursor.request_name:
        raise ProtocolError("inference watch changed request identity")
    if message.sequence != after + 1:
        raise ProtocolError("inference watch sequence is not contiguous")
    return message, inference_stream_pb2.InferenceStreamCursor(
        request_name=message.request_name,
        after_sequence=message.sequence,
        resume_token=message.resume_token,
    )


type InferenceCursor = inference_stream_pb2.InferenceStreamCursor | None
type InferenceWatchSpec = WatchSpec[
    inference_stream_pb2.InferenceStreamMessage,
    InferenceCursor,
]

_TERMINAL_UPDATES = frozenset({"final_result", "failure"})


def _inference_watch_spec(operation_name: str) -> InferenceWatchSpec:
    """Describe the inference watch to the shared resumable watcher."""

    name = required_text("inference operation name", operation_name)

    def build(cursor: InferenceCursor, remaining: float) -> Message:
        return _watch_request(name, cursor, remaining)

    def accept(
        raw: Message,
        cursor: InferenceCursor,
    ) -> tuple[inference_stream_pb2.InferenceStreamMessage, InferenceCursor, bool]:
        response = cast(inference_service_pb2.WatchInferenceResponse, raw)
        message, advanced = _accept_message(response, cursor)
        terminal = message.WhichOneof("update") in _TERMINAL_UPDATES
        return message, advanced, terminal

    return WatchSpec(
        method=WATCH_INFERENCE,
        build_request=build,
        accept=accept,
        closed_error=lambda: UnavailableError(
            "inference watch closed before terminal durable truth",
            retryable=True,
        ),
        timeout_error=lambda: TimeoutError("inference watch deadline expired"),
        cancelled_error=lambda: CancelledError("inference watch was cancelled"),
    )


class Inference(WithRawResponse):
    def __init__(self, invoker: SyncInvoker) -> None:
        self._invoker = invoker

    def submit(
        self,
        request: inference_request_pb2.InferenceRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        value, call = _materialize_request(self._invoker, request, options)
        response = cast(
            inference_service_pb2.SubmitInferenceResponse,
            self._invoker.unary(
                SUBMIT_INFERENCE,
                inference_service_pb2.SubmitInferenceRequest(inference_request=value),
                call=call,
                retry_safe=True,
            ),
        )
        operation = required_response_message(
            response, "operation", operation_pb2.Operation, label="inference submit"
        )
        required_text("operation id", operation.operation_id)
        return operation

    def get_request(
        self, name: str, *, options: CallOptions | None = None
    ) -> inference_request_pb2.InferenceRequest:
        call = prepare_call(
            options,
            default_timeout=self._invoker.config.default_timeout,
            require_idempotency=False,
        )
        response = cast(
            inference_service_pb2.GetInferenceRequestResponse,
            self._invoker.unary(
                GET_INFERENCE_REQUEST,
                inference_service_pb2.GetInferenceRequestRequest(
                    name=required_text("inference request name", name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "inference_request",
            inference_request_pb2.InferenceRequest,
            label="inference get request",
        )

    def get_result(
        self, operation_name: str, *, options: CallOptions | None = None
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        call = prepare_call(
            options,
            default_timeout=self._invoker.config.default_timeout,
            require_idempotency=False,
        )
        response = cast(
            inference_service_pb2.GetInferenceResultResponse,
            self._invoker.unary(
                GET_INFERENCE_RESULT,
                inference_service_pb2.GetInferenceResultRequest(
                    operation_name=required_text("inference operation name", operation_name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return (
            required_response_message(
                response,
                "result",
                inference_result_pb2.InferenceResult,
                label="inference get result",
            ),
            required_response_message(
                response,
                "operation",
                operation_pb2.Operation,
                label="inference get result",
            ),
        )

    def commit_result(
        self,
        command: inference_service_pb2.CommitInferenceResultRequest,
        *,
        options: CallOptions | None = None,
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        value, call = _materialize_commit(self._invoker, command, options)
        response = cast(
            inference_service_pb2.CommitInferenceResultResponse,
            self._invoker.unary(COMMIT_INFERENCE_RESULT, value, call=call, retry_safe=True),
        )
        return (
            required_response_message(
                response,
                "result",
                inference_result_pb2.InferenceResult,
                label="inference commit",
            ),
            required_response_message(
                response,
                "operation",
                operation_pb2.Operation,
                label="inference commit",
            ),
        )

    @streaming_method
    def watch(
        self,
        operation_name: str,
        *,
        cursor: inference_stream_pb2.InferenceStreamCursor | None = None,
        timeout: float = 300.0,
        cancellation: threading.Event | None = None,
        options: CallOptions | None = None,
    ) -> WatchStream[inference_stream_pb2.InferenceStreamMessage, InferenceCursor]:
        """Follow one inference stream, resuming from its durable cursor."""

        if cancellation is not None and cancellation.is_set():
            raise CancelledError("inference watch was cancelled")
        base, total = watch_budget(timeout, options)
        return WatchStream(
            self._invoker,
            _inference_watch_spec(operation_name),
            cursor=_copied_cursor(cursor),
            call=base,
            total=total,
            cancellation=cancellation,
        )

    def wait(
        self,
        operation_name: str,
        *,
        cursor: inference_stream_pb2.InferenceStreamCursor | None = None,
        timeout: float = 300.0,
        options: CallOptions | None = None,
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        for message in self.watch(operation_name, cursor=cursor, timeout=timeout, options=options):
            if message.WhichOneof("update") == "failure":
                raise ProtocolError("inference watch reported durable failure")
            if message.WhichOneof("update") == "final_result":
                return self.get_result(operation_name, options=options)
        raise ProtocolError("inference watch ended before terminal truth")


class AsyncInference(AsyncWithRawResponse):
    def __init__(self, invoker: AsyncInvoker) -> None:
        self._invoker = invoker

    async def submit(
        self,
        request: inference_request_pb2.InferenceRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        value, call = _materialize_request(self._invoker, request, options)
        response = cast(
            inference_service_pb2.SubmitInferenceResponse,
            await self._invoker.unary(
                SUBMIT_INFERENCE,
                inference_service_pb2.SubmitInferenceRequest(inference_request=value),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response, "operation", operation_pb2.Operation, label="inference submit"
        )

    async def get_request(
        self, name: str, *, options: CallOptions | None = None
    ) -> inference_request_pb2.InferenceRequest:
        call = prepare_call(
            options,
            default_timeout=self._invoker.config.default_timeout,
            require_idempotency=False,
        )
        response = cast(
            inference_service_pb2.GetInferenceRequestResponse,
            await self._invoker.unary(
                GET_INFERENCE_REQUEST,
                inference_service_pb2.GetInferenceRequestRequest(
                    name=required_text("inference request name", name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "inference_request",
            inference_request_pb2.InferenceRequest,
            label="inference get request",
        )

    async def get_result(
        self, operation_name: str, *, options: CallOptions | None = None
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        call = prepare_call(
            options,
            default_timeout=self._invoker.config.default_timeout,
            require_idempotency=False,
        )
        response = cast(
            inference_service_pb2.GetInferenceResultResponse,
            await self._invoker.unary(
                GET_INFERENCE_RESULT,
                inference_service_pb2.GetInferenceResultRequest(
                    operation_name=required_text("inference operation name", operation_name)
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return (
            required_response_message(
                response,
                "result",
                inference_result_pb2.InferenceResult,
                label="inference get result",
            ),
            required_response_message(
                response,
                "operation",
                operation_pb2.Operation,
                label="inference get result",
            ),
        )

    async def commit_result(
        self,
        command: inference_service_pb2.CommitInferenceResultRequest,
        *,
        options: CallOptions | None = None,
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        value, call = _materialize_commit(self._invoker, command, options)
        response = cast(
            inference_service_pb2.CommitInferenceResultResponse,
            await self._invoker.unary(COMMIT_INFERENCE_RESULT, value, call=call, retry_safe=True),
        )
        return (
            required_response_message(
                response,
                "result",
                inference_result_pb2.InferenceResult,
                label="inference commit",
            ),
            required_response_message(
                response,
                "operation",
                operation_pb2.Operation,
                label="inference commit",
            ),
        )

    @streaming_method
    def watch(
        self,
        operation_name: str,
        *,
        cursor: inference_stream_pb2.InferenceStreamCursor | None = None,
        timeout: float = 300.0,
        cancellation: asyncio.Event | None = None,
        options: CallOptions | None = None,
    ) -> AsyncWatchStream[inference_stream_pb2.InferenceStreamMessage, InferenceCursor]:
        """Follow one inference stream, resuming from its durable cursor."""

        if cancellation is not None and cancellation.is_set():
            raise CancelledError("inference watch was cancelled")
        base, total = watch_budget(timeout, options)
        return AsyncWatchStream(
            self._invoker,
            _inference_watch_spec(operation_name),
            cursor=_copied_cursor(cursor),
            call=base,
            total=total,
            cancellation=cancellation,
        )

    async def wait(
        self,
        operation_name: str,
        *,
        cursor: inference_stream_pb2.InferenceStreamCursor | None = None,
        timeout: float = 300.0,
        options: CallOptions | None = None,
    ) -> tuple[inference_result_pb2.InferenceResult, operation_pb2.Operation]:
        async for message in self.watch(
            operation_name, cursor=cursor, timeout=timeout, options=options
        ):
            if message.WhichOneof("update") == "failure":
                raise ProtocolError("inference watch reported durable failure")
            if message.WhichOneof("update") == "final_result":
                return await self.get_result(operation_name, options=options)
        raise ProtocolError("inference watch ended before terminal truth")

File: sdks/go/mindclade/inference.go
package mindclade

import (
	"context"
	"strings"

	"google.golang.org/protobuf/types/known/timestamppb"

	inferencev1 "github.com/mindclade/mindclade/protocols/generated/go/inference/v1"
	internalinferencev1 "github.com/mindclade/mindclade/protocols/generated/go/internalrpc/inference/v1"
	operationv1 "github.com/mindclade/mindclade/protocols/generated/go/operation/v1"
)

// InferenceService is a thin ergonomic façade over the generated internal
// InferenceService client. Every accepted and returned domain value remains an
// authoritative generated protobuf message.
type InferenceService struct {
	client    *Client
	transport internalinferencev1.InferenceServiceClient
}

// Submit freezes authenticated command context and submits generated inference
// intent. The caller retains ownership of request; the SDK clones before
// materializing transport identity and the canonical request digest.
func (service *InferenceService) Submit(ctx context.Context, request *inferencev1.InferenceRequest, options ...RequestOption) (*operationv1.Operation, error) {
	if request == nil || !validResourceIdentifier(request.GetName()) {
		return nil, &Error{Code: CodeInvalidArgument, Message: "generated inference request with a valid name is required"}
	}
	if service == nil || service.client == nil || service.transport == nil {
		return nil, &Error{Code: CodeFailedPrecondition, Message: "inference service is not configured"}
	}
	value := cloneGenerated(request)
	commandKey := ""
	if value.GetContext() != nil {
		commandKey = value.GetContext().GetIdempotencyKey()
	}
	callContext, metadata, cancel, err := service.client.mutationContext(ctx, commandKey, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	value.TenantId = service.client.config.TenantID
	value.ProjectId = service.client.config.ProjectID
	value.Context = nil
	digest, err := deterministicDigest(value)
	if err != nil {
		return nil, err
	}
	value.Context = commandContext(service.client.config, callContext, metadata, digest)
	response, err := service.transport.SubmitInference(callContext, &internalinferencev1.SubmitInferenceRequest{InferenceRequest: value})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetOperation() == nil {
		return nil, &Error{Code: CodeDataLoss, Message: "inference service returned no durable operation"}
	}
	return cloneGenerated(response.GetOperation()), nil
}

// GetRequest returns the immutable generated execution intent.
func (service *InferenceService) GetRequest(ctx context.Context, name string, options ...RequestOption) (*inferencev1.InferenceRequest, error) {
	if !validResourceIdentifier(name) {
		return nil, &Error{Code: CodeInvalidArgument, Message: "valid inference request name is required"}
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.GetInferenceRequest(callContext, &internalinferencev1.GetInferenceRequestRequest{Name: name})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetInferenceRequest() == nil {
		return nil, &Error{Code: CodeDataLoss, Message: "inference service returned no request"}
	}
	return cloneGenerated(response.GetInferenceRequest()), nil
}

// GetResult returns immutable terminal truth and the durable generated
// operation used to authorize that read.
func (service *InferenceService) GetResult(ctx context.Context, operationName string, options ...RequestOption) (*inferencev1.InferenceResult, *operationv1.Operation, error) {
	if !validResourceIdentifier(operationName) {
		return nil, nil, &Error{Code: CodeInvalidArgument, Message: "valid inference operation name is required"}
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, nil, err
	}
	defer cancel()
	response, err := service.transport.GetInferenceResult(callContext, &internalinferencev1.GetInferenceResultRequest{OperationName: operationName})
	if err != nil {
		return nil, nil, normalizeError(err)
	}
	if response.GetResult() == nil || response.GetOperation() == nil {
		return nil, nil, &Error{Code: CodeDataLoss, Message: "inference result response omitted result or operation"}
	}
	return cloneGenerated(response.GetResult()), cloneGenerated(response.GetOperation()), nil
}

// CommitResult sends a generated fenced terminal command. Authenticated
// identity and the canonical command digest are always rematerialized.
func (service *InferenceService) CommitResult(ctx context.Context, command *internalinferencev1.CommitInferenceResultRequest, options ...RequestOption) (*inferencev1.InferenceResult, *operationv1.Operation, error) {
	if command == nil || command.GetInferenceRequest() == nil || command.GetFence() == nil || command.GetResult() == nil || strings.TrimSpace(command.GetRequestDigest()) == "" {
		return nil, nil, &Error{Code: CodeInvalidArgument, Message: "complete generated inference result command is required"}
	}
	value := cloneGenerated(command)
	commandKey := ""
	if value.GetContext() != nil {
		commandKey = value.GetContext().GetIdempotencyKey()
	}
	callContext, metadata, cancel, err := service.client.mutationContext(ctx, commandKey, options...)
	if err != nil {
		return nil, nil, err
	}
	defer cancel()
	value.Context = nil
	digest, err := deterministicDigest(value)
	if err != nil {
		return nil, nil, err
	}
	value.Context = commandContext(service.client.config, callContext, metadata, digest)
	response, err := service.transport.CommitInferenceResult(callContext, value)
	if err != nil {
		return nil, nil, normalizeError(err)
	}
	if response.GetResult() == nil || response.GetOperation() == nil {
		return nil, nil, &Error{Code: CodeDataLoss, Message: "inference commit response omitted result or operation"}
	}
	return cloneGenerated(response.GetResult()), cloneGenerated(response.GetOperation()), nil
}

// InferenceWatcher is the inference spelling of the shared resumable watcher.
// It yields the generated stream message; its resume cursor is the generated
// InferenceStreamCursor the server issued with the last accepted message.
type InferenceWatcher = StreamWatcher[*inferencev1.InferenceStreamMessage, *inferencev1.InferenceStreamCursor]

// Watch streams generated inference updates, resuming from cursor. Passing the
// cursor a previous watcher reported through Cursor resumes exactly where that
// watcher stopped.
func (service *InferenceService) Watch(ctx context.Context, operationName string, cursor *inferencev1.InferenceStreamCursor, options ...RequestOption) (*InferenceWatcher, error) {
	if service == nil || service.client == nil || service.transport == nil || ctx == nil || !validResourceIdentifier(operationName) {
		return nil, &Error{Code: CodeInvalidArgument, Message: "context and valid inference operation name are required"}
	}
	watchContext, cancel, err := service.client.longRunningContext(ctx, options...)
	if err != nil {
		return nil, err
	}
	watcher, err := newStreamWatcher(watchContext, cancel, service.client.config, cloneGenerated(cursor), service.watchPolicy(operationName))
	if err != nil {
		cancel()
		return nil, err
	}
	return watcher, nil
}

// ResumeWatch continues a stream from a cursor a previous process persisted.
func (service *InferenceService) ResumeWatch(ctx context.Context, operationName string, cursor *inferencev1.InferenceStreamCursor, options ...RequestOption) (*InferenceWatcher, error) {
	return service.Watch(ctx, operationName, cursor, options...)
}

// watchPolicy keeps every inference-specific rule the previous hand-written
// watcher enforced: a complete message, a heartbeat bound to the last durable
// cursor, stable request identity, and a strictly contiguous sequence.
func (service *InferenceService) watchPolicy(operationName string) watchPolicy[*inferencev1.InferenceStreamMessage, *inferencev1.InferenceStreamCursor] {
	return watchPolicy[*inferencev1.InferenceStreamMessage, *inferencev1.InferenceStreamCursor]{
		open: func(ctx context.Context, cursor *inferencev1.InferenceStreamCursor) (func() (*inferencev1.InferenceStreamMessage, error), error) {
			request := &internalinferencev1.WatchInferenceRequest{OperationName: operationName, Cursor: cloneGenerated(cursor)}
			if deadline, ok := ctx.Deadline(); ok {
				request.Deadline = timestamppb.New(deadline)
			}
			stream, err := service.transport.WatchInference(ctx, request)
			if err != nil {
				return nil, normalizeError(err)
			}
			return func() (*inferencev1.InferenceStreamMessage, error) {
				response, receiveErr := stream.Recv()
				if receiveErr != nil {
					return nil, receiveErr
				}
				return response.GetMessage(), nil
			}, nil
		},
		accept: func(cursor *inferencev1.InferenceStreamCursor, message *inferencev1.InferenceStreamMessage) (*inferencev1.InferenceStreamCursor, bool, error) {
			if message == nil || message.GetRequestName() == "" || message.GetResumeToken() == "" || message.GetSequence() == 0 {
				return cursor, false, &Error{Code: CodeDataLoss, Message: "inference watch returned an incomplete message"}
			}
			if message.GetHeartbeat() != nil {
				if cursor == nil || message.GetRequestName() != cursor.GetRequestName() || message.GetSequence() != cursor.GetAfterSequence() || message.GetResumeToken() != cursor.GetResumeToken() {
					return cursor, false, &Error{Code: CodeDataLoss, Message: "inference heartbeat was not bound to the last durable cursor"}
				}
				return cursor, false, nil
			}
			after := uint64(0)
			if cursor != nil {
				after = cursor.GetAfterSequence()
				if cursor.GetRequestName() != message.GetRequestName() {
					return cursor, false, &Error{Code: CodeDataLoss, Message: "inference watch changed request identity"}
				}
			}
			if message.GetSequence() != after+1 {
				return cursor, false, &Error{Code: CodeDataLoss, Message: "inference watch sequence is not contiguous"}
			}
			next := &inferencev1.InferenceStreamCursor{RequestName: message.GetRequestName(), AfterSequence: message.GetSequence(), ResumeToken: message.GetResumeToken()}
			return next, message.GetFinalResult() != nil || message.GetFailure() != nil, nil
		},
		snapshot: cloneGenerated[*inferencev1.InferenceStreamCursor],
	}
}

// Wait consumes resumable typed updates until terminal truth is durable, then
// returns the authoritative generated result and operation.
func (service *InferenceService) Wait(ctx context.Context, operationName string, cursor *inferencev1.InferenceStreamCursor, options ...RequestOption) (*inferencev1.InferenceResult, *operationv1.Operation, error) {
	watcher, err := service.Watch(ctx, operationName, cursor, options...)
	if err != nil {
		return nil, nil, err
	}
	defer func() { _ = watcher.Close() }()
	for {
		message, receiveErr := watcher.Recv()
		if receiveErr != nil {
			return nil, nil, receiveErr
		}
		if message.GetFailure() != nil {
			return nil, nil, &Error{Code: CodeFailedPrecondition, Message: "inference watch reported durable failure"}
		}
		if message.GetFinalResult() != nil {
			return service.GetResult(ctx, operationName, options...)
		}
	}
}
